usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2545,6 +2607,959 @@ CK_RV cca_load_config_file(STDLL_TokData_t *tokdata, char *conf_name)
+
+    return ret;
+}
+
+/*
+ * Save the current firmware wrapping key verification pattern in the tokdata:
+ * create a dummy test key, transform it into a protected key, and store the wk
+ * verification pattern in the tokdata.
+ */
+static CK_RV ccatok_pkey_get_firmware_wkvp(STDLL_TokData_t *tokdata)
+{
+    struct cca_private_data *cca_data = tokdata->private_data;
+    CK_ATTRIBUTE *pkey_attr = NULL, *sec_attr = NULL;
+    CK_RV ret;
+    long return_code = 0, reason_code = 0;
+    long exit_data_len = 0, clear_key_bit_length = 0;
+    unsigned char exit_data[4];
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE];
+    long rule_array_count;
+    unsigned char *clear_key_value = NULL;
+    long key_name_length = 0;
+    unsigned char key_name[CCA_KEY_ID_SIZE] = { 0, };
+    long user_data_length = 0;
+    unsigned char user_data[64] = { 0, };
+    long token_data_length = 0;
+    long verb_data_length = 0;
+    unsigned char verb_data[64] = { 0, };
+    unsigned char the_key_token[CCA_KEY_TOKEN_SIZE] = { 0, };
+    long the_key_token_length = sizeof(the_key_token);
+    unsigned char key_type_1[8] = { ' ', };
+    unsigned char key_type_2[8] = { ' ', };
+    long key_name_1_length = 0;
+    long key_name_2_length = 0;
+    long user_data_1_length = 0;
+    long user_data_2_length = 0;
+    long kek_identifier_1_length = 0;
+    long kek_identifier_2_length = 0;
+    long the_2nd_key_token_length = 0;
+    unsigned int min_card_version;
+
+    /* Check CCA host library version: XPRTCPAC requires min 7.0.0 */
+    if (cca_data->cca_lib_version.ver < 7) {
+        TRACE_ERROR("CCA host lib is %d.%d.%d, but pkey support requires min 7.0.0\n",
+                    cca_data->cca_lib_version.ver,cca_data->cca_lib_version.rel,
+                    cca_data->cca_lib_version.mod);
+        OCK_SYSLOG(LOG_ERR, "CCA host lib is %d.%d.%d, but pkey support requires min 7.0.0\n",
+                   cca_data->cca_lib_version.ver,cca_data->cca_lib_version.rel,
+                   cca_data->cca_lib_version.mod);
+        return CKR_FUNCTION_NOT_SUPPORTED;
+    }
+
+    /* Read min card version from CCA private data. Needs a read lock. */
+    if (pthread_rwlock_rdlock(&cca_data->min_card_version_rwlock) != 0) {
+        TRACE_ERROR("CCA min_card_version RD-Lock failed.\n");
+        return CKR_CANT_LOCK;
+    }
+    min_card_version = cca_data->min_card_version.ver;
+    if (pthread_rwlock_unlock(&cca_data->min_card_version_rwlock) != 0) {
+        TRACE_ERROR("CCA min_card_version RD-Unlock failed.\n");
+        return CKR_CANT_LOCK;
+    }
+
+    /* Check minimum card version, XPRTCPAC requires min CEX7C */
+    if (min_card_version < 7) {
+        TRACE_ERROR("Minimum card version must be CEX7C, but we only have %d.%d.%d\n",
+                    cca_data->min_card_version.ver,cca_data->min_card_version.rel,
+                    cca_data->min_card_version.mod);
+        OCK_SYSLOG(LOG_ERR, "Minimum card version must be CEX7C, but we only have %d.%d.%d\n",
```

Comment:
Please make this a LOG_WARN, this is really not an error situation. It just means we can't use PKEY.

Reply:
p.s. will also change the TRACE_ERROR to TRACE_WARNING then